### DISCUSSION TITLE
How to deallocate a pointer in C?

### DISCUSSION THREAD

**Discussion by ghost** [community/community#105305] [Programming Help] [answered]
### Body

```cpp
#include <stdio.h>
#include <string>
extern "C"{
    struct RCS{
        std::string*memory[3];
        void writeToMemory(std::string *item){
            for(int i = 0; i < 3; i++){
                if(*memory[i] == ""){
                    *memory[i] = *item;
                }else{
                    *memory[i] = "";
                    *memory[i] = *item;
                }
            }
        }
        void clearMemory(){
            for(int i = 0; i < 3; i++){
                if(*memory[i] != ""){
                    *memory[i] = "";
                }else{
                    printf("Cannot clear the memory because it is already cleared!\n");
                }
            }
        }
        void deallocateStringMemory(){
            for(int i = 0; i < 4; i++){
                memory[i] = nullptr;
                delete[] memory;
            }
        }
    };
    struct RID{
        double *doubleMemory[3];
        int *intMemory[3];
        void writeToDoubleMemory(int index, double value){
            if(*doubleMemory[index] == 0){
                *doubleMemory[index] = value;
            }else{
                *doubleMemory[index] = 0;
                *doubleMemory[index] = value;
            }
        }
        void clearDoubleMemory(int index){
            if(*doubleMemory[index] > 0){
                *doubleMemory[index] = 0;
            }else{
                printf("Cannot clear memory because it is already empty!\n");
            }
        }
        void writeToIntMemory(int index, int value){
            if(*intMemory[index] == 0){
                *intMemory[index] = value;
            }else{
                printf("Cannot write to memory because it is full!\n");
            }
        }
        void clearIntMemory(int index){
            if(*intMemory[index] > 0){
                *intMemory[index] = 0;
            }else{
                printf("Cannot clear memory because it is already empty \n");
            }
        }
        void deallocateIntMemory(){
            for(int i = 0; i < 4; i++){
                intMemory[i] = nullptr;
                delete[] intMemory;
            }
        }
        void deallocateDoubleMemory(){
            for(int i = 0; i < 4; i++){
                doubleMemory[i] = nullptr;
                delete[] doubleMemory;
            }
        }
    };
}
```

### Guidelines

- [x] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by thenaserov** [comment]
free(your_pointer_var);
Even if your pointer is already freed you will be good to go!

**Reply by ghost** [reply]
How does the heap and free work? I am new to low level programming.

**Comment by davea42** [comment, marked as answer]
Please be careful.  This is a 'generic' reply about malllc/free and does
not discuss the specifics of your code.

```
char *x = (char *) malloc(20);
if (x) {
    free(x);
    x = 0;
    free(x);
}
```

In the above delete 'memory';

is valid, but if you omit the line of code zeroing the pointer it is a major bug,
a disaster which valgrind or  gcc -fsanitise or other validators (coverity scan)
will detect and report.

So.   On doing a free, zero the pointer.
If you don't do the x=0; in the above the result
is errors or unpredictable crashes.

free(0) is always valid and does nothing (per the C Standard).

For C++:
Similarly delete(x) does not zero x, so deleting it twice 
without zeroing x in between is ... very bad.

The heap is an (invisible to the coder) large block of memory
(the size is not meaningfully available to you, though 
since this is running a real program on a real/virtual machine
with non-infinite memory  the
size will matter if you ask for too much).

Malloc(calloc, realloc)  use the memory and build their own data structures
(invisible to you)  and record each malloc/calloc/realloc so that free(x) can
look up x in those hidden data structures and remove some memory
from the set of allocated blocks.

There have been many implementations of  malloc() etc and free over time
and it's performance (in time and space) is markedly better  now than it was
years ago.  And of course memory size on modern computers is ... large

It is not necessary to understand the algorithms involved, but all this
is visible in the source to libc (at least for Linux) so you can study it.

I hope this  makes sense.
Corrections welcome.
